core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -234,6 +226,59 @@ impl ZkSyncStateKeeper {
         Err(Error::Canceled)
     }
 
+    async fn load_protocol_upgrade_tx(
+        &mut self,
+        pending_miniblocks: &[MiniblockExecutionData],
+        protocol_version: ProtocolVersionId,
+        first_batch_in_shared_bridge: bool,
+    ) -> Result<Option<ProtocolUpgradeTx>, Error> {
+        let previous_batch_protocol_version =
+            self.io.load_previous_batch_version_id().await.unwrap();
+
+        let version_changed = protocol_version != previous_batch_protocol_version;
+        let mut protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
+
+        // After the Shared Bridge is integrated,
+        // there has to be a setChainId upgrade transaction after the chain genesis.
+        // It has to be the first transaction of the first batch.
+        // If it's not present, we have to wait for it to be picked up by the event watcher.
+        // The setChainId upgrade does not bump the protocol version, rather attaches an upgrade
+        // transaction to the genesis protocol version.
+        while first_batch_in_shared_bridge && protocol_upgrade_tx.is_none() {
+            tracing::info!("Waiting for setChainId upgrade transaction to be picked up");
+            if self.is_canceled() {
+                return Err(Error::Canceled);
+            }
+            tokio::time::sleep(Duration::from_secs(1)).await;
+            protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
+        }
+
+        let protocol_upgrade_tx =
+            if pending_miniblocks.is_empty() && (version_changed || first_batch_in_shared_bridge) {
+                tracing::info!("We have upgrade tx to be executed in empty miniblock");
+                protocol_upgrade_tx
+            } else if !pending_miniblocks.is_empty()
+                && (version_changed || first_batch_in_shared_bridge)
+            {
+                // Sanity check: if `txs_to_reexecute` is not empty and upgrade tx is present for this block
+                // then it must be the first one in `txs_to_reexecute`.
+                if protocol_upgrade_tx.is_some() {
+                    let first_tx_to_reexecute = &pending_miniblocks[0].txs[0];
+                    assert_eq!(
+                        first_tx_to_reexecute.tx_format(),
+                        TransactionType::ProtocolUpgradeTransaction
+                    )
```

Comment:
if we create assert like this - let's add more information here to make debugging easier. (for example, the transaction hash of the first transaction etc)